symfony/plugins/orangehrmAuthenticationPlugin/Service/ResetPasswordService.php

```diff
@@ -0,0 +1,334 @@
+                $this->getLogger()->error('employee work email was not set');
+            } else {
+                return $user;
+            }
+        }
+        $this->getLogger()->error('user account was deleted');
+        return null;
+    }
+
+    /**
+     * @param string $resetCode
+     * @return User|null
+     */
+    public function validateUrl(string $resetCode): ?User
+    {
+        $userNameMetaData = $this->extractPasswordResetMetaData($resetCode);
+        if (count($userNameMetaData) > 0) {
+            $username = $userNameMetaData[0];
+            $resetPassword = $this->getResetPasswordDao()->getResetPasswordLogByResetCode($resetCode);
+            if ($resetPassword instanceof ResetPassword) {
+                $expDay = $this->hasPasswordResetRequestNotExpired($resetPassword);
+                if ($expDay > 0) {
+                    $this->getLogger()->error('not valid URL');
+                    return null;
+                }
+                $user = $this->getUserService()->getSystemUserDao()->getUserByUserName($username);
+                return $this->validateUser($user, true);
+            }
+            return null;
+        }
+        $this->getLogger()->error('Invalid reset code');
+        return null;
+    }
+
+    /**
+     * @param ResetPassword $resetPassword
+     * @return ResetPassword|null
+     */
+    public function saveResetPasswordLog(ResetPassword $resetPassword): ?ResetPassword
+    {
+        return $this->getResetPasswordDao()->saveResetPassword($resetPassword);
+    }
+
+    /**
+     * @param User $user
+     * @return bool
+     */
+    public function logPasswordResetRequest(User $user): bool
+    {
+        $identifier = $user->getUserName();
+        $resetCode = $this->generatePasswordResetCode($identifier);
+        $resetPassword = new ResetPassword();
+        $resetPassword->setResetEmail($user->getEmployee()->getWorkEmail());
+        $date = $this->getDateTimeHelper()->getNow();
+        $resetPassword->setResetRequestDate($date);
+        $resetPassword->setResetCode($resetCode);
+        $emailSent = $this->sendPasswordResetCodeEmail($user->getEmployee(), $resetCode);
+        if (!$emailSent) {
+            $this->getLogger()->error('Password reset email could not be sent.');
+            return false;
+        }
+        $this->saveResetPasswordLog($resetPassword);
+        return true;
+    }
+
+    /**
+     * @param string $password
+     * @param string $userName
+     * @return bool
+     */
+    public function saveResetPassword(string $password, string $userName): bool
+    {
+        $user = $this->getUserService()->getSystemUserDao()->getUserByUserName($userName);
+        if ($this->validateUser($user, false) instanceof User) {
+            $hashPassword = $this->getUserService()->hashPassword($password);
+            return $this->getUserService()->getSystemUserDao()->updatePassword($user->getId(), $hashPassword);
+        }
+        return false;
+    }
```

Comment:
can't we reuse UserService methods.
Shouldn't provide username, password into method params. incase execution  failed stack trace may contain password. need to discuss